### PR TITLE
feat: add support for routers

### PR DESCRIPTION
**Pull request by LiLatee**
Hey! 👋 

This PR allows for testing widgets that depend on some kinds of Routers e.g. [go_router](https://pub.dev/packages/go_router). To be precise in one of the projects we use 
```dart
bool _isActive(BuildContext context) => GoRouter.of(context).location == NotificationsPage.routeToPush;
```
in order to check if we are currently on that specific page. Because there is no Router provided in the widget tree the test fails. It's just a single example, but I guess there can be more similar cases for different routers 🤔 

So basically this PR adds an option for using `MaterialApp.router` if someone needs it. Because using `MaterialApp.router` does not have a `child` param we are not able to wrap our widget that we want to test into some additional widgets as e.g. `Provider`. That's why I also added a `wrappers` param which wraps a whole `MaterialApp`.

Let me know what you think and what I should adjust to your requirements. 👌

### CODE REVIEW

File: lib/src/golden_test.dart [hunk 1]
```diff
@@ -86,71 +109,102 @@ void goldenTest({
...
+    ),
+    'If you want to use `MaterialApp.router` then provide all related fields. Otherwise provide none of them and provide `builder` parameter.',
+  );
+
+  final bool areAllRouterFieldsProvided = [
+    routerDelegate,
+    routeInformationProvider,
+    routeInformationParser
+  ].every((element) => element != null);
+  if (areAllRouterFieldsProvided) {
```
Comment (dmkasperski) [inline lib/src/golden_test.dart:203]: chore: please separate the logic for theming the widget with MaterialApp or MaterialApp.router. From the goldenTest call something like `_themedMaterialApp` but inside based on `areAllRouterFieldsProvided` call one of the 2 methods to create one of the MaterialApp constructors

File: lib/src/golden_test.dart [hunk 2]
```diff
@@ -53,10 +54,28 @@ import 'package:meta/meta.dart';
...
+///       child: child,
+///     ),
+///   );
+/// ```
+/// {@end-tool}
 @isTest
 void goldenTest({
   required String name,
-  required WidgetBuilder builder,
+  WidgetBuilder? builder,
```
Comment (dmkasperski) [inline lib/src/golden_test.dart:78]: chore: please keep builder as a required parameter, so that writing tests is as easy as possible, so that IDE can autocomplete params for you and you don't have to add them "manually"

Depending on what your needs for testing router, how similar are those cases, how many such cases in the project you have, how they differ from each other, you can 
1. create a separate method for tests where you use router
2. use the global config
3. pass the router with required `builder` and if the router is not null, create route inside router

it would be much simpler to suggest some solution if you could create some example when, why and how you test it.

The example you gave in description, where you check on what route you are (I guess you have some generic route creation) seems that it should be mocked (what I guess you still do with a wrapper) instead of creating actual router

**Review by dmkasperski — CHANGES_REQUESTED** [review]
Thank you for contribution! I'm looking forward to merge those changes, but would be great if you could provide some changes before 🙏
